Implement free text query

tim-preston opened the pull request:
We've been testing Prometheus and when we looked at weathermap integration we found your project.

It worked great for us and we used it with snmp interface metrics collected from Juniper, Extreme and Edgecore (running IPInfusion's OCNOS software) devices.

For what we want to use in production we'd be looking at putting large LAG bundles on our maps and so I added in free-text queries which allow us to use just about anything we have in the metrics as the datasource and works really well for us. In the examples I put in the readme I've showed aggregate traffic and busiest link queries.

We've been running this in the lab for a few weeks now with no issues and we'll be going live with this shortly so I think that this should be good enough to contribute back.

tim-preston committed:
Add free-test queries to the manual test script

tim-preston committed:
Add free-test queries to the daatsource script.

tim-preston committed:
Fix comments.

tim-preston committed:
Update README.

lodpp commented on README.md:
````diff
@@ -118,6 +134,24 @@ LINK NODE1-NODE2
+```
+LINK NODE1-NODE2
+    NODES NODE1 NODE2
+    TARGET max(irate(ifHCOutOctets{ifAlias=~".*ae232.*",instance="192.168.178.32"}[10m]) or irate(ifHCOutOctets{ifAlias=~".*ae232.*",instance="192.168.178.32"}[10m]))*8
````
I think that it's missing at the beginning the whole `prometheus:http:localhost:9090:` part to be
```
TARGET prometheus:http:localhost:9090:max(irate(ifHCOutOctets{ifAlias=~".*ae232.*",instance="192.168.178.32"}[10m]) or irate(ifHCOutOctets{ifAlias=~".*ae232.*",instance="192.168.178.32"}[10m]))*8
```

tim-preston replied:
Fixed

lodpp commented on README.md:
````diff
@@ -118,6 +134,24 @@ LINK NODE1-NODE2
+    TARGET max(irate(ifHCOutOctets{ifAlias=~".*ae232.*",instance="192.168.178.32"}[10m]) or irate(ifHCOutOctets{ifAlias=~".*ae232.*",instance="192.168.178.32"}[10m]))*8
+```
+These free-text PromQL queries can be very powerful.
+
````
You could add a friendly warning that depending on the query, it can be quite slow.
Depending on the query, number of queries and the how often you refresh the wmap(s), this can have an impact.
Maybe people that uses the datasource are probably aware of that already and we don't care though. Your call :)

tim-preston replied:
Added a quick warning. It can't hurt =)

lodpp commented on WeatherMapDataSource_prometheus.php:
```diff
@@ -29,11 +29,28 @@
+// prometheus:proto:remote_host:remote_port:free_text_query_in
+
+// This exmple shows how powerful this can be. It uses the max() with regex matching to poll the busiest 
+// link in a bundle on a given routyer, identified here by having 'ae232' in the interface descriotion 
```
typo: s/routyer/router/ and s/descriotion/description/

lodpp commented on WeatherMapDataSource_prometheus.php:
```diff
@@ -104,8 +123,8 @@ function ReadData($targetstring, &$map, &$item)
     
-            $in_query  = 'irate(' . $in_series  . '{ifName="' . $intf . '",instance="' . $instance . '"}[2m])*8';
-            $out_query = 'irate(' . $out_series . '{ifName="' . $intf . '",instance="' . $instance . '"}[2m])*8';
+            $in_query  = 'irate(' . $in_series  . '{ifName="' . $intf . '",instance="' . $instance . '"}[10m])*8';
```
Non-blocking:
Funny, I've moved to 20m on some devices that are slow as hell to be polled ( nexus7k + lots of fexes ) and was looking at making this tunable in TARGET ( with a default to some value )

tim-preston replied:
These changes were left over from initial lab testing where we currently only poll every 5m. I've reverted them since is was unintended churn.
Making this settable in TARGET for the fixed-form interface query would make sense.

lodpp commented on WeatherMapDataSource_prometheus_manualtest.php:
```diff
@@ -11,21 +11,27 @@
 
 
 //$prom_query = 'prometheus:http:localhost:9090:192.168.178.32:Gi0/13:ifHCInOctets:ifHCOutOctets';
+//$prom_query = 'prometheus:http:localhost:9090:sum(irate(ifHCOutOctets{ifAlias=~".*ae232.*",instance="192.168.178.32"}[10m]))*8:sum(irate(ifHCInOctets{ifAlias=~".*ae232.*",instance="192.168.178.32"}[10m]))*8';
```
FWIW the IN should appear first, to be in line with the in_query and out_query

tim-preston replied:
Fixed,

lodpp reviewed (changes requested):
Nice add :)
I was thinking of doing the same to graph other thing than network bw, so great to get it more flexible.
Minor things to modify otherwise LGTM

tim-preston committed:
Fix exmaple weathermap config.
Add a warning about overloading prometheus with expensive free-text queries.

tim-preston committed:
Fix typos in comments.
Revert changes to the range-ventor in the interface queries; unintentionally committed from lab testing.

tim-preston committed:
Fix order of in & out metrics in example free-text query.

lodpp commented:
thanks a lot for your contribution !
